src/com/amazon/corretto/crypto/provider/AesXtsSpi.java

```diff
@@ -0,0 +1,293 @@
+  }
+
+  private boolean checkKeyTweakEquality(final byte[] key, final byte[] tweak) {
+    for (int i = 0; i != TWEAK_SIZE_IN_BYTES; i++) {
+      if (packedTweakKey[i] != tweak[i]) return false;
+    }
+
+    for (int i = 0; i != KEY_SIZE_IN_BYTES; i++) {
+      if (packedTweakKey[i + TWEAK_SIZE_IN_BYTES] != key[i]) return false;
+    }
+
+    return true;
+  }
+
+  @Override
+  protected byte[] engineUpdate(byte[] input, int inputOffset, int inputLen) {
+    throw new UnsupportedOperationException(
+        "Multiple-part encryption or decryption is not supported. Consider using DoFinal.");
+  }
+
+  @Override
+  protected int engineUpdate(
+      final byte[] input,
+      final int inputOffset,
+      final int inputLen,
+      final byte[] output,
+      final int outputOffset) {
+    throw new UnsupportedOperationException(
+        "Multiple-part encryption or decryption is not supported. Consider using DoFinal.");
+  }
+
+  @Override
+  protected byte[] engineDoFinal(byte[] input, int inputOffset, int inputLen)
+      throws IllegalBlockSizeException, BadPaddingException {
+    final byte[] output = new byte[getExactOutputSize(inputLen)];
+    try {
+      engineDoFinal(input, inputOffset, inputLen, output, 0);
+    } catch (final ShortBufferException e) {
+      throw new AssertionError(e);
+    }
+    return output;
+  }
+
+  @Override
+  protected int engineDoFinal(
+      final byte[] input,
+      final int inputOffset,
+      final int inputLen,
+      final byte[] output,
+      final int outputOffset)
+      throws ShortBufferException, IllegalBlockSizeException, BadPaddingException {
+    cipherState.checkCipherStateForFinal();
+    Utils.checkArrayLimits(input, inputOffset, inputLen);
+    final int outputLen = output.length - outputOffset;
+    Utils.checkArrayLimits(output, outputOffset, outputLen);
+    final int actualOutputLen = getExactOutputSize(inputLen);
+    if (outputLen < actualOutputLen) {
+      throw new ShortBufferException("Output buffer is not large enough.");
+    }
+
+    if (Utils.outputClobbersInput(input, inputOffset, inputLen, output, outputOffset)) {
+      final byte[] tempOutput = engineDoFinal(input, inputOffset, inputLen);
+      System.arraycopy(tempOutput, 0, output, outputOffset, tempOutput.length);
+    } else {
+      if (cipherState == CipherState.ENCRYPT_INIT) {
+        encrypt(input, inputOffset, inputLen, output, outputOffset);
+      } else {
+        decrypt(input, inputOffset, inputLen, output, outputOffset);
+      }
+    }
+
+    cipherState = cipherState.nextStateAfterDoFinal();
+    return actualOutputLen;
+  }
+
+  private void encrypt(
+      byte[] input, int inputOffset, int inputLen, byte[] output, int outputOffset) {
+    if (input != output) {
+      enc(packedTweakKey, input, inputOffset, inputLen, output, outputOffset);
+    } else {
+      encSameBuffer(packedTweakKey, input, inputOffset, inputLen, outputOffset);
```

Comment:
It handles it correctly but in this way, we can reduce the number of `GetPrimitiveArrayCritical/ReleasePrimitiveArrayCritical` invocations by one since they are the same buffer.

Reply:
got it. are the `{Get,Release}PrimitiveArrayCritical` calls expensive? my understanding is that these just return a view on memory and [manage some lock state/object pinning](https://github.com/corretto/corretto-11/blob/f4cc07566cf589fac42893ed8d926d4fae23b087/src/hotspot/share/prims/jni.cpp#L3190), but don't do any allocation.